File: solutions/typescript/0068.ts
function fullJustify(words: string[], maxWidth: number): string[] {
    const result: string[] = [];
    let i = 0;

    while (i < words.length) {
        let lineLen = words[i].length;
        let j = i + 1;
        while (j < words.length && lineLen + 1 + words[j].length <= maxWidth) {
            lineLen += 1 + words[j].length;
            j++;
        }

        const lineWords = words.slice(i, j);
        const totalChars = lineWords.reduce((sum, w) => sum + w.length, 0);
        const gaps = lineWords.length - 1;
        const isLastLine = j === words.length;

        if (gaps === 0 || isLastLine) {
            let line = lineWords.join(' ');
            line += ' '.repeat(maxWidth - line.length);
            result.push(line);
        } else {
            const totalSpaces = maxWidth - totalChars;
            const spacePerGap = Math.floor(totalSpaces / gaps);
            let extraSpaces = totalSpaces % gaps;
            let line = '';
            for (let k = 0; k < lineWords.length; k++) {
                line += lineWords[k];
                if (k < gaps) {
                    line += ' '.repeat(spacePerGap + (extraSpaces > 0 ? 1 : 0));
                    extraSpaces--;
                }
            }
            result.push(line);
        }

        i = j;
    }

    return result;
}

export default fullJustify;

File: solutions/rust/0068.rs
impl Solution {
    pub fn full_justify(words: Vec<String>, max_width: i32) -> Vec<String> {
        let max_width = max_width as usize;
        let mut result: Vec<String> = Vec::new();
        let mut i = 0;

        while i < words.len() {
            let mut line_len = words[i].len();
            let mut j = i + 1;
            while j < words.len() && line_len + 1 + words[j].len() <= max_width {
                line_len += 1 + words[j].len();
                j += 1;
            }

            let line_words = &words[i..j];
            let total_chars: usize = line_words.iter().map(|w| w.len()).sum();
            let gaps = line_words.len() - 1;
            let is_last_line = j == words.len();

            if gaps == 0 || is_last_line {
                let mut line = line_words.join(" ");
                let padding = max_width - line.len();
                line.push_str(&" ".repeat(padding));
                result.push(line);
            } else {
                let total_spaces = max_width - total_chars;
                let space_per_gap = total_spaces / gaps;
                let mut extra_spaces = total_spaces % gaps;
                let mut line = String::new();
                for k in 0..line_words.len() {
                    line.push_str(&line_words[k]);
                    if k < gaps {
                        let spaces = space_per_gap + if extra_spaces > 0 { 1 } else { 0 };
                        if extra_spaces > 0 {
                            extra_spaces -= 1;
                        }
                        line.push_str(&" ".repeat(spaces));
                    }
                }
                result.push(line);
            }

            i = j;
        }

        result
    }
}
